llvm/lib/Target/SPIRV/SPIRVOperands.td

```diff
@@ -0,0 +1,1515 @@
+// capabilities. 
+//===----------------------------------------------------------------------===//
+
+def Capability : GenericEnum, Operand<i32> {
+  let FilterClass = "Capability";
+  let NameField = "Name";
+  let ValueField = "Value";
+  let PrintMethod = !strconcat("printSymbolicOperand<OperandCategory::", FilterClass, "Operand>");
+}
+
+class Capability<string name, bits<32> value> {
+  string Name = name;
+  bits<32> Value = value;
+}
+
+multiclass CapabilityOperand<bits<32> value, bits<32> minVersion, bits<32> maxVersion, list<Extension> reqExtensions, list<Capability> reqCapabilities> {
+  def NAME : Capability<NAME, value>;
+  defm : SymbolicOperandWithRequirments<CapabilityOperand, value, NAME, minVersion, maxVersion, reqExtensions, reqCapabilities>;
+}
+
+defm Matrix : CapabilityOperand<0, 0, 0, [], []>;
+defm Shader : CapabilityOperand<1, 0, 0, [], [Matrix]>;
+defm Geometry : CapabilityOperand<2, 0, 0, [], [Shader]>;
+defm Tessellation : CapabilityOperand<3, 0, 0, [], [Shader]>;
+defm Addresses : CapabilityOperand<4, 0, 0, [], []>;
+defm Linkage : CapabilityOperand<5, 0, 0, [], []>;
+defm Kernel : CapabilityOperand<6, 0, 0, [], []>;
+defm Vector16 : CapabilityOperand<7, 0, 0, [], [Kernel]>;
+defm Float16Buffer : CapabilityOperand<8, 0, 0, [], [Kernel]>;
+defm Float16 : CapabilityOperand<9, 0, 0, [], []>;
+defm Float64 : CapabilityOperand<10, 0, 0, [], []>;
+defm Int64 : CapabilityOperand<11, 0, 0, [], []>;
+defm Int64Atomics : CapabilityOperand<12, 0, 0, [], [Int64]>;
+defm ImageBasic : CapabilityOperand<13, 0, 0, [], [Kernel]>;
+defm ImageReadWrite : CapabilityOperand<14, 0, 0, [], [ImageBasic]>;
+defm ImageMipmap : CapabilityOperand<15, 0, 0, [], [ImageBasic]>;
+defm Pipes : CapabilityOperand<17, 0, 0, [], [Kernel]>;
+defm Groups : CapabilityOperand<18, 0, 0, [], []>;
+defm DeviceEnqueue : CapabilityOperand<19, 0, 0, [], []>;
+defm LiteralSampler : CapabilityOperand<20, 0, 0, [], [Kernel]>;
+defm AtomicStorage : CapabilityOperand<21, 0, 0, [], [Shader]>;
+defm Int16 : CapabilityOperand<22, 0, 0, [], []>;
+defm TessellationPointSize : CapabilityOperand<23, 0, 0, [], [Tessellation]>;
+defm GeometryPointSize : CapabilityOperand<24, 0, 0, [], [Geometry]>;
+defm ImageGatherExtended : CapabilityOperand<25, 0, 0, [], [Shader]>;
+defm StorageImageMultisample : CapabilityOperand<27, 0, 0, [], [Shader]>;
+defm UniformBufferArrayDynamicIndexing : CapabilityOperand<28, 0, 0, [], [Shader]>;
+defm SampledImageArrayDymnamicIndexing : CapabilityOperand<29, 0, 0, [], [Shader]>;
+defm ClipDistance : CapabilityOperand<32, 0, 0, [], [Shader]>;
+defm CullDistance : CapabilityOperand<33, 0, 0, [], [Shader]>;
+//defm ImageCubeArray : CapabilityOperand<34, 0, 0, [], [SampledCubeArray]>;
+defm SampleRateShading : CapabilityOperand<35, 0, 0, [], [Shader]>;
+//defm ImageRect : CapabilityOperand<36, 0, 0, [], [SampledRect]>;
+defm SampledRect : CapabilityOperand<37, 0, 0, [], [Shader]>;
+
+defm ImageRect : CapabilityOperand<36, 0, 0, [], [SampledRect]>;
```

Comment:
Shouldn't we just uncomment this on line 301?

Reply:
Right, thanks! Forgot about these lines! The commented lines should be removed - I will update the patch. The definitions for _ImageRect_ must be lower (at line 304) since _ImageRect_ requires _SampledRect_ capability, which is defined just above at line 302. For the very same reason _ImageCubeArray_ is defined lower.

I don't think there is any way to "forward declare" TableGen entries -- that would resolve this problem. Though this issue only happens for the two _ImageRect_ and _ImageCubeArray_ definitions.